### PR TITLE
Fix deadlock when enrollment invalid while writing logs

### LINKED ISSUE #219: Deadlock when writing logs requires reenrollment
Consider the following stacktrace from @marpaia's launcher process that is not checking in:

```
goroutine 38 [semacquire, 654 minutes]:
sync.runtime_SemacquireMutex(0xc4201dc17c, 0xc420182200)
	/usr/local/go/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc4201dc178)
	/usr/local/go/src/sync/mutex.go:134 +0xee
github.com/kolide/launcher/vendor/github.com/boltdb/bolt.(*DB).beginRWTx(0xc4201dc000, 0x0, 0x0, 0x0)
	/Users/marpaia/go/src/github.com/kolide/launcher/vendor/github.com/boltdb/bolt/db.go:512 +0x65
github.com/kolide/launcher/vendor/github.com/boltdb/bolt.(*DB).Begin(0xc4201dc000, 0xc400000001, 0xc42005c080, 0xb5, 0x142ca3c)
	/Users/marpaia/go/src/github.com/kolide/launcher/vendor/github.com/boltdb/bolt/db.go:461 +0x38
github.com/kolide/launcher/vendor/github.com/boltdb/bolt.(*DB).Update(0xc4201dc000, 0x15d25a0, 0x0, 0x0)
	/Users/marpaia/go/src/github.com/kolide/launcher/vendor/github.com/boltdb/bolt/db.go:582 +0x46
github.com/kolide/launcher/osquery.(*Extension).RequireReenroll(0xc4201c8140, 0x1873860, 0xc42001a050)
	/Users/marpaia/go/src/github.com/kolide/launcher/osquery/extension.go:250 +0x97
github.com/kolide/launcher/osquery.(*Extension).writeLogsWithReenroll(0xc4201c8140, 0x1873860, 0xc42001a050, 0x4, 0xc42024a000, 0x30, 0x40, 0x1, 0xc42001f970, 0xc42001f900)
	/Users/marpaia/go/src/github.com/kolide/launcher/osquery/extension.go:459 +0x103
github.com/kolide/launcher/osquery.(*Extension).writeBufferedLogsForType.func1(0xc42021a000, 0x15d26a8, 0xc42021a000)
	/Users/marpaia/go/src/github.com/kolide/launcher/osquery/extension.go:431 +0x382
github.com/kolide/launcher/vendor/github.com/boltdb/bolt.(*DB).Update(0xc4201dc000, 0xc420220de0, 0x0, 0x0)
	/Users/marpaia/go/src/github.com/kolide/launcher/vendor/github.com/boltdb/bolt/db.go:598 +0x9a
github.com/kolide/launcher/osquery.(*Extension).writeBufferedLogsForType(0xc4201c8140, 0x4, 0x104ce6b, 0xc420154380)
	/Users/marpaia/go/src/github.com/kolide/launcher/osquery/extension.go:400 +0xba
github.com/kolide/launcher/osquery.(*Extension).writeAndPurgeLogs(0xc4201c8140)
	/Users/marpaia/go/src/github.com/kolide/launcher/osquery/extension.go:356 +0x8e
github.com/kolide/launcher/osquery.(*Extension).writeLogsLoopRunner(0xc4201c8140)
	/Users/marpaia/go/src/github.com/kolide/launcher/osquery/extension.go:380 +0xd1
created by github.com/kolide/launcher/osquery.(*Extension).Start
	/Users/marpaia/go/src/github.com/kolide/launcher/osquery/extension.go:143 +0x5f
```

Note that `writeBufferedLogsForType` opens a Bolt update transaction (https://github.com/kolide/launcher/blob/master/osquery/extension.go#L400), then calls `writeLogsWithReenroll`. In `writeLogsWithReenroll`, if the enrollment is invalid a reenrollment is attempted (https://github.com/kolide/launcher/blob/master/osquery/extension.go#L454-L459). This reenrollment attempts to open an Update transaction (https://github.com/kolide/launcher/blob/master/osquery/extension.go#L250) and deadlocks.

The log writing code should be updated to not attempt to recursively acquire an Update, and other code should be audited to look for similar problems.

### PR DESCRIPTION
**Pull request by zwass**
Fixes #219

### CODE REVIEW

File: osquery/extension.go [hunk 1]
```diff
@@ -397,10 +397,13 @@ func (e *Extension) writeBufferedLogsForType(typ logger.LogType) error {
 	if err != nil {
 		return err
 	}
-	err = e.db.Update(func(tx *bolt.Tx) error {
+
+	// Collect up logs to be sent
+	logs := []string{}
```
Comment (groob) [inline osquery/extension.go]: I would prefer this starts with `var logs []string`  and takes advantage of Go's ability to append to a declared slice.

File: osquery/extension.go [hunk 2]
```diff
@@ -397,10 +397,13 @@ func (e *Extension) writeBufferedLogsForType(typ logger.LogType) error {
...
+	logIDs := [][]byte{}
```
Comment (groob) [inline osquery/extension.go]: same as above.

File: osquery/extension.go [hunk 3]
```diff
@@ -417,30 +420,38 @@ func (e *Extension) writeBufferedLogsForType(typ logger.LogType) error {
...
-			// deleted.
-			return errors.Wrap(err, "writing logs")
-		}
+	err = e.writeLogsWithReenroll(context.Background(), typ, logs, true)
+	if err != nil {
+		return errors.Wrap(err, "writing logs")
+	}
 
+	// Delete logs that were successfully sent
+	err = e.db.Update(func(tx *bolt.Tx) error {
```
Comment (groob) [inline osquery/extension.go]: can you document how logs are keyed in boltdb?

Comment (zwass) [reply osquery/extension.go]: Added that here: https://github.com/kolide/launcher/pull/220/files#diff-cfe59979dfc77c83ca9ee6b3dabc765cR548

**Pull request merged**